lib/internal/http2/compat.js

```diff
@@ -201,7 +207,7 @@ class Http2ServerRequest extends Readable {
   }
 
   get socket() {
-    return this.stream.session.socket;
+    return this[kStream].session.socket;
```

Comment:
`this[kStream]` can be nulled, shouldn't we protect this?

Reply:
Yeah, I agree. It wasn't protected before so I wasn't sure if there was a reason for it or not.

On that note, hapi uses this to set their internal processing flag and since our stream reference isn't accessible after 'finish', it throws. Should we maybe store a reference to the session directly and not null it on finish? It could lead to some garbage collection issues but on the flipside it's the only issue I found with hapi compatibility.

Reply:
@apapirovski I don't think what HAPI is actually doing in HTTP1 could be ported 1-1 to HTTP2. I've added this to be compatibile but for all intent and purposes this should not be used: the socket is controlled by the C++/C layers. 

@jasnell what do you think?

Reply:
Yeah, it's a bit weird because they don't actually seem to use the socket based on my quick review of the code — maybe they did in the past. Now they just store a flag on it. When I built a version that retained the socket reference then my test site worked fine with h2 on hapi. But I agree with you re: not using it.

I guess it's all about how much we're willing to compromise for compatibility.

Reply:
Just a thought but if we do something like `process.nextTick(() => (this[kStream] = undefined));` then we get the best of both worlds. It passes all our tests, makes hapi work and gets rid of the stream reference after the user finishes processing the 'finish' event. That's probably the most h1 compatible solution.